### PR TITLE
Migrate some provider to ASP.NET Core 2.0

### PR DESCRIPTION
**Pull request by kinosang**
Port some providers to the new ASP.NET Core 2.0 authentication stack.

* [x] ArcGIS
* [x] Asana
* [x] Autodesk
* [x] BattleNet
* [x] Discord
* [x] Dropbox
* [x] Imgur
* [x] Paypal
* [x] Reddit
* [x] Twitch
* [x] Vimeo
* [x] Weibo
* [x] Weixin
* [x] WordPress
* [x] Yahoo
* [x] Yandex

### CODE REVIEW

**Comment by huww98** [comment]
I want to add the QQ provider for ASP.NET Core 2.0 based on #160. Maybe I will create a PR today.

**Comment by kevinchalet** [comment]
Awesome, thanks! Could you please squash the intermediate commits to only have the `Migrate XYZ` ones?

**Comment by kinosang** [comment]
@PinpointTownes I'll do it asap.

**Comment by kinosang** [comment]
@PinpointTownes all done.

File: src/AspNet.Security.OAuth.ArcGIS/ArcGISAuthenticationHandler.cs [hunk 1]
```diff
@@ -55,20 +58,15 @@ public ArcGISAuthenticationHandler([NotNull] HttpClient client)
                                 /* Code: */ error.Value<string>("code"),
                                 /* Message: */ error.Value<string>("message"));
 
-                throw new InvalidOperationException("An error occurred while retrieving the user profile.");
+                throw new HttpRequestException("An error occurred while retrieving user information.");
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.ArcGIS/ArcGISAuthenticationHandler.cs]: Any particular reason to change the error message?

File: src/AspNet.Security.OAuth.Paypal/PaypalAuthenticationHandler.cs [hunk 2]
```diff
@@ -46,19 +51,14 @@ public PaypalAuthenticationHandler([NotNull] HttpClient client)
...
-                    .AddOptionalClaim(ClaimTypes.Email, PaypalAuthenticationHelper.GetEmail(payload), Options.ClaimsIssuer);
-
             var principal = new ClaimsPrincipal(identity);
-            var ticket = new AuthenticationTicket(principal, properties, Options.AuthenticationScheme);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload);
+            context.RunClaimActions(payload);
 
-            var context = new OAuthCreatingTicketContext(ticket, Context, Options, Backchannel, tokens, payload);
-            await Options.Events.CreatingTicket(context);
+            identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, payload.Value<string>("user_id")?.Split('/')?.LastOrDefault(), ClaimValueTypes.String, Options.ClaimsIssuer));
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Paypal/PaypalAuthenticationHandler.cs]: FYI, you can use `ClaimActions.MapCustomJson` to add claims contained in nested JSON objects. E.g: https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNetCore.Authentication.MicrosoftAccount/MicrosoftAccountOptions.cs#L32

Any chance we could use it everywhere we can?

File: src/AspNet.Security.OAuth.Vimeo/VimeoAuthenticationHandler.cs [hunk 3]
```diff
@@ -46,17 +51,14 @@ public VimeoAuthenticationHandler([NotNull] HttpClient client)
...
-                    .AddOptionalClaim("urn:vimeo:profileurl", VimeoAuthenticationHelper.GetProfileUrl(payload), Options.ClaimsIssuer);
-
             var principal = new ClaimsPrincipal(identity);
-            var ticket = new AuthenticationTicket(principal, properties, Options.AuthenticationScheme);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload);
+            context.RunClaimActions(payload);
 
-            var context = new OAuthCreatingTicketContext(ticket, Context, Options, Backchannel, tokens, payload);
-            await Options.Events.CreatingTicket(context);
+            identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, payload.Value<string>("uri")?.Split('/')?.LastOrDefault(), ClaimValueTypes.String, Options.ClaimsIssuer));
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Vimeo/VimeoAuthenticationHandler.cs]: Same remark (I see this pattern in a few places, that would be cool to move to the claims actions API).

File: src/AspNet.Security.OAuth.Weibo/WeiboAuthenticationHandler.cs [hunk 4]
```diff
@@ -53,25 +58,56 @@ public WeiboAuthenticationHandler([NotNull] HttpClient client)
...
-                    .AddOptionalClaim("urn:weibo:location", WeiboAuthenticationHelper.GetLocation(payload), Options.ClaimsIssuer);            
-
             var principal = new ClaimsPrincipal(identity);
-            var ticket = new AuthenticationTicket(principal, properties, Options.AuthenticationScheme);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload);
+            context.RunClaimActions(payload);
 
-            var context = new OAuthCreatingTicketContext(ticket, Context, Options, Backchannel, tokens);
-            await Options.Events.CreatingTicket(context);
+            // When the email address is not public, retrieve it from
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Weibo/WeiboAuthenticationHandler.cs]: I'm not against adding that, but not as part of this PR. Please remove it temporarily from the associated commit and open a new ticket to bring it back in both the ASP.NET Core 1.0 and 2.0 versions.

**Review by kevinchalet — CHANGES_REQUESTED** [review]

**Comment by kevinchalet** [comment]
@kinosang thank you very much. I added a few remarks but overall it looks great. Please ping me when they are addressed and I'll merge your PR ASAP.

**Comment by kinosang** [comment]
@PinpointTownes done, waiting for CI testing.

**Comment by kinosang** [comment]
syn @PinpointTownes

**Comment by kinosang** [comment]
ping @PinpointTownes

**Review by kevinchalet — APPROVED** [review]

**Comment by kevinchalet** [comment]
@kinosang sorry for the long delay, I was busy with tons of other things.

I just merged your PR. Thank you very much for your effort, it's much appreciated! :clap:

**Pull request closed without merging**